### PR TITLE
Start previous epoch cluster consensus components after epoch boundary

### PR DESCRIPTION
**Pull request by jordanschalm**
This PR addresses an edge case where cluster consensus components for a just-ended epoch would not be initialized upon startup of the `epochmgr` engine. When the engine starts up within the first `tx_expiry` blocks of an epoch, cluster consensus should be re-started for the previous epoch, to service still-valid transactions referencing blocks within the previous epoch.

### Localnet Test
Collection Node 3 is stopped, then restarted. Both its current and previous epoch consensus instances continue making progress after the restart.

Screen Shot 2023-02-16 at 4 21 34 PM
Screen Shot 2023-02-16 at 4 21 50 PM
Screen Shot 2023-02-16 at 4 22 01 PM

### CODE REVIEW

File: engine/collection/epochmgr/engine_test.go [hunk 1]
```diff
@@ -265,7 +280,66 @@ func (suite *Suite) TestRestartInSetupPhase() {
...
+	suite.header.Height = prevEpochFinalHeight + 1
+	suite.heights.On("OnHeight", prevEpochFinalHeight+flow.DefaultTransactionExpiry+1, mock.Anything)
+
+	suite.StartEngine()
+	// previous epoch components should have been started
+	suite.AssertEpochStarted(suite.counter - 1)
+	suite.AssertEpochStarted(suite.counter)
+}
+
+// TestStartAfterEpochBoundary tests starting the engine shortly after an epoch transition.
```
Comment (AlexHentschel) [inline engine/collection/epochmgr/engine_test.go:305]:
```suggestion
// TestStartAfterEpochBoundary_BeyondTxExpiry tests starting the engine shortly after an epoch transition.
```

File: engine/collection/epochmgr/engine_test.go [hunk 2]
```diff
@@ -139,13 +125,33 @@ func (suite *Suite) SetupTest() {
 			},
 			func(epoch realprotocol.Epoch) component.Component { return suite.ComponentsForEpoch(epoch).messageHub },
 			func(epoch realprotocol.Epoch) error { return nil },
-		)
+		).Maybe()
+}
+
+func (suite *Suite) SetupTest() {
+
```
Comment (AlexHentschel) [inline engine/collection/epochmgr/engine_test.go:132]:
```suggestion
```

File: module/builder/collection/builder.go [hunk 3]
```diff
@@ -40,6 +40,10 @@ type Builder struct {
 	log            zerolog.Logger
 }
 
+// TODO: #6435
+//   - pass in epoch (minimally counter, preferably cluster chain ID as well)
+//   - check candidate reference blocks by view (cheap, but need to get whole header each time - cheap if header in cache)
```
Comment (AlexHentschel) [inline module/builder/collection/builder.go:45]:
```suggestion
//   - check candidate reference blocks by view (need to get whole header each time - cheap if header in cache)
```

**Review by AlexHentschel — APPROVED** [review]
very nice. Thanks for the detailed tests and the great documentation.

File: engine/collection/epochmgr/engine.go [hunk 4]
```diff
@@ -120,7 +120,7 @@ func (e *Engine) Start(ctx irrecoverable.SignalerContext) {
 	}
 
 	// (4) start epoch-scoped components:
-	// set up epoch-scoped epoch managed by this engine for the current epoch
+	// (a) set up epoch-scoped epoch managed by this engine for the current epoch
 	components, err := e.createEpochComponents(currentEpoch)
 	if err != nil {
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply engine/collection/epochmgr/engine.go:144]: Nice catch - changed to start current/previous epoch clusters independently in https://github.com/onflow/flow-go/pull/3921/commits/aa5d3cfc4f0ff25cbde0712d5aba51f32813bd48

**Review by durkmurder — APPROVED** [review]
🎸 nice tests, thanks for upgrading to new mocks

**Comment by jordanschalm** [comment]
bors merge

**Pull request merged**